resolver.go

```diff
@@ -500,6 +500,12 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 			break
 		}
 		if resp == nil {
+			logrus.Debugf("[resolver] failed to resolve through external DNS servers")
+			resp = new(dns.Msg)
+			resp.SetRcode(query, dns.RcodeServerFailure)
+			if err = w.WriteMsg(resp); err != nil {
```

Comment:
Can you skip the `err = w.WriteMsg(resp)` + error checking/logs + return since it's already done later down at the end of the function. So just set up the resp and have the `err = w.WriteMsg(resp)` down below take care of it.